src/fs_indexer.cc

```diff
@@ -25,14 +25,26 @@ fs_indexer::fs_indexer(code_searcher *cs,
 fs_indexer::~fs_indexer() {
 }
 
-void fs_indexer::read_file(const string& path) {
+void fs_indexer::read_file(const fs::path& path) {
     ifstream in(path.c_str(), ios::in);
-    string relpath(mismatch(path.begin(), path.end(), repopath_.begin()).first,
-                   path.end());
-    cs_->index_file(tree_, relpath, StringPiece(static_cast<stringstream const&>(stringstream() << in.rdbuf()).str().c_str(), fs::file_size(path)));
+    fs::path relpath = fs::relative(path, repopath_);
+    cs_->index_file(tree_, relpath.string(), StringPiece(static_cast<stringstream const&>(stringstream() << in.rdbuf()).str().c_str(), fs::file_size(path)));
 }
 
-void fs_indexer::walk(const string& path) {
+void fs_indexer::walk_contents_file(const fs::path& contents_file_path) {
+    ifstream contents_file(contents_file_path.c_str(), ios::in);
+    if (!contents_file.is_open()) {
+        throw std::ifstream::failure("Unable to open contents file for reading");
```

Comment:
This all looks wonderful! One last little tweak and I'll merge: could you include the path in the error message? Otherwise, in a big log file where lots of repositories were indexed, it can be hard to track down which contents file went missing. I'm only asking for this at this moment because I'm pretty sure this is the first error I'm likely to make as I try deploying this new approach — is getting the deploy order wrong or something and one of the content files isn't there!